### PR TITLE
Treat net_receive and post_event as API methods.

### LINKED ISSUE #1425: membrane potential not available in NET_RECEIVE
The membrane potential is not available in `NET_RECEIVE` via the magic variable `v`:

```
diff --git a/mechanisms/default/expsyn.mod b/mechanisms/default/expsyn.mod
index 44da3697..625dec95 100644
--- a/mechanisms/default/expsyn.mod
+++ b/mechanisms/default/expsyn.mod
@@ -33,6 +33,6 @@ DERIVATIVE state {
 } NET_RECEIVE(weight) {
-    g = g + weight
+    g = g + weight + v
 }
```

Compiling with `modcc expsyn.mod -t cpu` yields:

``` 
error: expsyn.mod (line 36,col 22) 
  the symbol 'v' refers to an external quantity and is unavailable in a function or procedure
error: expsyn.mod (line 36,col 20) 
  the symbol 'v' refers to an external quantity and is unavailable in a function or procedure
error: expsyn.mod (line 36,col 7) 
  the symbol 'v' refers to an external quantity and is unavailable in a function or procedure
error: expsyn.mod (line 36,col 5) 
  the symbol 'v' refers to an external quantity and is unavailable in a function or procedure
error: expsyn.mod (line 35,col 1) 
  the symbol 'v' refers to an external quantity and is unavailable in a function or procedure
error: error   expsyn.mod:(line 1,col 1) There were 5 errors in the semantic analysis
```

### PR DESCRIPTION
**Pull request by noraabiakar**
- Give `net_receive` and `post_event` access to the API scope.
- Create `APIMethods` from `net_receive` and `post_event`, and use them for generating the C++ and CUDA code. 

Fixes #1425

### CODE REVIEW

**Comment by noraabiakar** [comment]
bors try

File: modcc/printer/gpuprinter.cpp [hunk 1]
```diff
@@ -445,7 +445,7 @@ void emit_api_body_cu(std::ostream& out, APIMethod* e, bool is_point_proc) {
     }
 
     if (!body->statements().empty()) {
-        out << "int tid_ = threadIdx.x + blockDim.x*blockIdx.x;\n";
+        cv_loop && out << "int tid_ = threadIdx.x + blockDim.x*blockIdx.x;\n";
```
Comment (bcumming) [inline modcc/printer/gpuprinter.cpp:448]: The generated kernel for `apply_events` requires the thread index, e.g
```C++
__global__
void apply_events(int mech_id_, mechanism_gpu_expsyn_pp_ params_, deliverable_event_stream_state events) {
    auto tid_ = threadIdx.x + blockDim.x*blockIdx.x;
    auto const ncell_ = events.n;

    if(tid_<ncell_) { /* .. */}
}  
```
~Are these being printed elsewhere?~

On further reading, I found where the `tid_` is calculated.

Should the `auto tid_ = ...` etc. be hoisted outside the `emit_api_body_cu`?

**Review by bcumming — CHANGES_REQUESTED** [review]
Looks good.

There is one possible fix: hoisting api-method specific loop logic outside the loop printer to obviate the need for the `cv_loop` parameter.

File: modcc/printer/gpuprinter.cpp (hunk 1 above)
Comment (noraabiakar) [reply modcc/printer/gpuprinter.cpp]: Yes, I think that's a good idea. I just pushed the changes.

**Comment by noraabiakar** [comment]
bors try

**Review by bcumming — APPROVED** [review]

**Review by halfflat — APPROVED** [review]
Great! This is less messy than I feared :)

**Pull request merged**